federationsender/queue/queue.go

```diff
@@ -200,15 +235,44 @@ func (oqs *OutgoingQueues) SendEDU(
 		)
 	}
 
-	// Remove our own server from the list of destinations.
-	destinations = filterAndDedupeDests(oqs.origin, destinations)
+	// Deduplicate destinations and remove the origin from the list of
+	// destinations just to be sure.
+	destmap := map[gomatrixserverlib.ServerName]struct{}{}
+	for _, d := range destinations {
+		destmap[d] = struct{}{}
+	}
+	delete(destmap, oqs.origin)
 
-	if len(destinations) > 0 {
-		log.WithFields(log.Fields{
-			"destinations": destinations, "edu_type": e.Type,
-		}).Info("Sending EDU event")
+	// There is absolutely no guarantee that the EDU will have a room_id
+	// field, as it is not required by the spec. However, if it *does*
+	// (e.g. typing notifications) then we should try to make sure we don't
+	// bother sending them to servers that are prohibited by the server
+	// ACLs.
+	var header struct {
+		RoomID string `json:"room_id"`
+	}
+	if err := json.Unmarshal(e.Content, &header); err == nil {
```

Comment:
I imagine `gjson` would be faster than unmarshalling into a temp struct, it also reads nicer.

Reply:
Good point. I'll update it to use gjson.